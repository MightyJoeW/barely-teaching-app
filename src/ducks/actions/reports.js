import uuid from "uuid";
import database from "../../firebase/firebase";

// ADD_REPORT
export const addReport = report => ({
  type: "ADD_REPORT",
  report
});

export const startAddReport = (reportData = {}) => {
  return dispatch => {
    const {
      student_name = "",
      note = "",
      period = 0,
      createdAt = 0
    } = reportData;
    const report = { student_name, note, period, createdAt };

    return database
      .ref("reports")
      .push(report)
      .then(ref => {
        dispatch(
          addReport({
            id: ref.key,
            ...report
          })
        );
      });
  };
};

// REMOVE_REPORT
export const removeReport = ({ id } = {}) => ({
  type: "REMOVE_REPORT",
  id
});

export const startRemoveReport = ({ id } = {}) => {
  return dispatch => {
    return database
      .ref(`reports/${id}`)
      .remove()
      .then(() => {
        dispatch(removeReport({ id }));
      });
  };
};

// EDIT_REPORT
export const editReport = (id, updates) => ({
  type: "EDIT_REPORT",
  id,
  updates
});

export const startEditReport = (id, updates) => {
  return dispatch => {
    return database
      .ref(`reports/${id}`)
      .update(updates)
      .then(() => {
        dispatch(editReport(id, updates));
      });
  };
};

// SET_REPORTS
export const setReports = reports => ({
  type: "SET_REPORTS",
  reports
});

export const startSetReports = () => {
  return dispatch => {
    return database
      .ref("reports")
      .once("value")
      .then(snapshot => {
        const reports = [];

        snapshot.forEach(childSnapshot => {
          reports.push({
            id: childSnapshot.key,
            ...childSnapshot.val()
          });
        });

        dispatch(setReports(reports));
      });
  };
};
