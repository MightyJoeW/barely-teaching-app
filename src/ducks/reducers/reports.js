// Reports Reducer

const reportsReducerDefaultState = [];

export default (state = reportsReducerDefaultState, action) => {
  switch (action.type) {
    case "ADD_REPORT":
      return [...state, action.report];
    case "REMOVE_REPORT":
      return state.filter(({ id }) => id !== action.id);
    case "EDIT_REPORT":
      return state.map(report => {
        if (report.id === action.id) {
          return { ...report, ...action.updates };
        } else {
          return report;
        }
      });
    default:
      return state;
  }
};
