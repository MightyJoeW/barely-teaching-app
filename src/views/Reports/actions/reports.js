import uuid from "uuid";

// ADD_REPORT
export const addReport = (
  { student_name = "", note = "", period = 0, createdAt = 0 } = {}
) => ({
  type: "ADD_REPORT",
  report: {
    id: uuid(),
    student_name,
    note,
    period,
    createdAt
  }
});

// REMOVE_REPORT
export const removeReport = ({ id } = {}) => ({
  type: "REMOVE_REPORT",
  id
});

// EDIT_REPORT
export const editReport = (id, updates) => ({
  type: "EDIT_REPORT",
  id,
  updates
});
