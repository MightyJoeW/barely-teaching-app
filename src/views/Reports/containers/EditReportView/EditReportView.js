import React, { Component } from "react";
import { connect } from "react-redux";
import ReportForm from "../../components/ReportForm";
import {
  startEditReport,
  startRemoveReport
} from "../../../../ducks/actions/reports";

export class EditReportView extends Component {
  onSubmit = report => {
    this.props.startEditReport(this.props.report.id, report);
    this.props.history.push("/reports");
  };
  onRemove = () => {
    this.props.startRemoveReport({ id: this.props.report.id });
    this.props.history.push("/reports");
  };
  render() {
    return (
      <div>
        <div className="">
          <div className="content-container">
            <h1 className="reports-header__title"> Edit Report </h1>
          </div>
        </div>
        <div className="content-container">
          <ReportForm report={this.props.report} onSubmit={this.onSubmit} />
          <button
            className="login-button button--secondary"
            onClick={this.onRemove}
          >
            Remove Report
          </button>
        </div>
      </div>
    );
  }
}

const mapStateToProps = (state, props) => ({
  report: state.reports.find(report => report.id === props.match.params.id)
});

const mapDispatchToProps = (dispatch, props) => ({
  startEditReport: (id, report) => dispatch(startEditReport(id, report)),
  startRemoveReport: data => dispatch(startRemoveReport(data))
});

export default connect(mapStateToProps, mapDispatchToProps)(EditReportView);
