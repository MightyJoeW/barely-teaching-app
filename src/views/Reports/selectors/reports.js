import moment from "moment";

// Get visible reports
export default (reports, { text, sortBy, startDate, endDate }) => {
  return reports
    .filter(report => {
      const createdAtMoment = moment(report.createdAt);
      const startDateMatch = startDate
        ? startDate.isSameOrBefore(createdAtMoment, "day")
        : true;
      const endDateMatch = endDate
        ? endDate.isSameOrAfter(createdAtMoment, "day")
        : true;
      const textMatch = report.student_name
        .toLowerCase()
        .includes(text.toLowerCase());

      return startDateMatch && endDateMatch && textMatch;
    })
    .sort((a, b) => {
      if (sortBy === "date") {
        return a.createdAt < b.createdAt ? 1 : -1;
      } else if (sortBy === "period") {
        return a.period < b.period ? 1 : -1;
      }
    });
};
