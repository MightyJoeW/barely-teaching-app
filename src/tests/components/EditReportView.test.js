import React from "react";
import { shallow } from "enzyme";
import reports from "../fixtures/reports";
import { EditReportView } from "../../views/Reports/containers/EditReportView/EditReportView";

let editReport, removeReport, history, wrapper;

beforeEach(() => {
  editReport = jest.fn();
  removeReport = jest.fn();
  history = { push: jest.fn() };
  wrapper = shallow(
    <EditReportView
      editReport={editReport}
      removeReport={removeReport}
      history={history}
      report={reports[2]}
    />
  );
});

test("should render EditReportView", () => {
  expect(wrapper).toMatchSnapshot();
});

test("should handle editReport", () => {
  wrapper.find("ReportForm").prop("onSubmit")(reports[2]);
  expect(history.push).toHaveBeenLastCalledWith("/reports");
  expect(editReport).toHaveBeenLastCalledWith(reports[2].id, reports[2]);
});

test("should handle removeReport", () => {
  wrapper.find("button").simulate("click");
  expect(history.push).toHaveBeenLastCalledWith("/reports");
  expect(removeReport).toHaveBeenLastCalledWith({
    id: reports[2].id
  });
});
