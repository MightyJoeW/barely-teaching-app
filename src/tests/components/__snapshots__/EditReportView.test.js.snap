// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render EditReportView 1`] = `
<div>
  <div
    className="reports-header"
  >
    <div
      className="content-container"
    >
      <h1
        className="reports-header__title"
      >
         Edit Report 
      </h1>
    </div>
  </div>
  <div
    className="content-container"
  >
    <ReportForm
      onSubmit={[Function]}
      report={
        Object {
          "createdAt": 345600000,
          "id": "3",
          "note": "",
          "period": 3,
          "student_name": "Fred",
        }
      }
    />
    <button
      className="login-button button--secondary"
      onClick={[Function]}
    >
      Remove Report
    </button>
  </div>
</div>
`;
